apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracerBuilder.java

```diff
@@ -153,6 +155,21 @@ public ElasticApmTracer build() {
         return tracer;
     }
 
+    /**
+     * A noop method aimed for working around the JULI deadlock reported at https://github.com/elastic/apm-agent-java/issues/954.
+     * The main thread creates an {@link ApmServerClient} and then starts the Tracer, which start the {@link LifecycleListener}s.
+     * One of the listeners is the JMX metric tracker, that initializes the JUL mechanism. Concurrently, another thread is
+     * doing healthcheck with the APM server (kicked from the {@link ApmServerClient}), and this may cause a deadlock.
+     * Calling this method before both, should prevent this deadlock.
+     */
+    private void initializeJavaUtilLogging() {
+        try {
+            java.util.logging.Logger.getLogger("noop");
```

Comment:
I'm pretty sure that this will fail the JBoss integration tests. JBoss needs to be able to set the `java.util.logging.manager` system property to override the default `LogManager`.

See also

https://github.com/elastic/apm-agent-java/blob/96b7d60e4edcb39242ab5d7922111d52fd9c1023/apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java#L96-L144

I assume because the health check runs in a background thread, JBoss has enough time to init the log manager. Maybe we can make the above more code more generic and fire the start lifecycle event only after `java.util.logging.manager` has been set if running on JBoss and fire immediately otherwise.

Reply:
That's odd, the fact that they it does get into this deadlock means that the server healthcheck is done quickly enough. But we can
>  make the above more code more generic and fire the start lifecycle event only after java.util.logging.manager has been set if running on JBoss and fire immediately otherwise

to be on the safe side.

Reply:
OK, there are a lot of hidden subtleties that produce misleading behaviours here. 

1. Why didn't we see this before with the healthcheck? This is the relevant thread stack taken from the deadlock thread dump:
```
"apm-server-healthcheck":
        at java.util.logging.LogManager.ensureLogManagerInitialized(LogManager.java:317)
        - waiting to lock <0x0000000755bb5ef0> (a org.apache.juli.ClassLoaderLogManager)
        at java.util.logging.LogManager.getLogManager(LogManager.java:378)
        at java.util.logging.Logger.getPlatformLogger(Logger.java:572)
        at java.util.logging.LoggingProxyImpl.getLogger(LoggingProxyImpl.java:41)
        at sun.util.logging.LoggingSupport.getLogger(LoggingSupport.java:100)
        at sun.util.logging.PlatformLogger$JavaLoggerProxy.<init>(PlatformLogger.java:602)
        at sun.util.logging.PlatformLogger$JavaLoggerProxy.<init>(PlatformLogger.java:597)
        at sun.util.logging.PlatformLogger.<init>(PlatformLogger.java:239)
        at sun.util.logging.PlatformLogger.getLogger(PlatformLogger.java:198)
        - locked <0x0000000756becbe0> (a java.lang.Class for sun.util.logging.PlatformLogger)
        at sun.net.www.protocol.http.HttpURLConnection.<clinit>(HttpURLConnection.java:431)
        at sun.net.www.protocol.http.Handler.openConnection(Handler.java:62)
        at sun.net.www.protocol.http.Handler.openConnection(Handler.java:57)
        at java.net.URL.openConnection(URL.java:979)
        at co.elastic.apm.agent.report.ApmServerClient.startRequestToUrl(ApmServerClient.java:125)
        at co.elastic.apm.agent.report.ApmServerClient.executeForAllUrls(ApmServerClient.java:236)
        at co.elastic.apm.agent.report.ApmServerHealthChecker.call(ApmServerHealthChecker.java:69)
        at co.elastic.apm.agent.report.ApmServerHealthChecker.call(ApmServerHealthChecker.java:47)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```
The `sun.net.www.protocol.http.HttpURLConnection` static initializer calls `sun.util.logging.PlatformLogger.getLogger()` explicitly, which in turn creates a new `PlatformLogger` instance. This constructor uses the `JavaLoggerProxy` (resulting in the rest of the stack above) only if some `java.util.logging.config` system properties were previously set (see what I needed to do in my [code snippet](https://github.com/elastic/apm-agent-java/pull/1203#issue-423224248) above in order to be able to reproduce the deadlock). If such are not set, it will instantiate a `PlatformLogger.DefaultLoggerProxy` instead, which will not continue with the JUL initialization chain, thus not interfering with WildFly's Log Manager obsession.
2. So why don't the WildFly/JBoss integration tests fail when calling `Logger.getLogger` explicitly? Because of shading! The call I added is replaced with `co.elastic.apm.agent.logging.JulBridgeLogger.getLogger` 🤦 

Going back to the drawing board...